fix(compiler-cli): generate valid TS 5.6 type checking code

crisbeto opened the pull request:
Currently in some scenarios the compiler generates code like `null as any ? foo : bar` which will be invalid with [an upcoming TypeScript change](https://devblogs.microsoft.com/typescript/announcing-typescript-5-6-beta/#disallowed-nullish-and-truthy-checks). These changes switch to generating `0 as any` which is exempt from the new behavior.

**Note:** I'm not starting the work to fully get us on TS 5.6 until the 18.2 release comes out, but this change is necessary to unblock an internal team.

crisbeto commented:
[Passing TGP](http://OCL:660728753:BASE:660765227:1723112782108:b9c672fe)

devversion reviewed (approved):
LGTM, great comment!

crisbeto committed:
fix(compiler-cli): generate valid TS 5.6 type checking code

Currently in some scenarios the compiler generates code like `null as any ? foo : bar` which will be invalid with [an upcoming TypeScript change](https://devblogs.microsoft.com/typescript/announcing-typescript-5-6-beta/#disallowed-nullish-and-truthy-checks). These changes switch to generating `0 as any` which is exempt from the change.

**Note:** I'm not starting the work to fully get us on TS 5.6 until the 18.2 release comes out, but this change is necessary to unblock an internal team.

AndrewKushnir commented:
This PR was merged into the repository by commit ca55b3d4547296d722596bb196045c814cf5fec0.

The changes were merged into the following branches: main, 18.1.x, 18.2.x
